### ISSUE TITLE
(nucleus): Infinite loop broken pipe in logs

### ISSUE THREAD

**Issue by fiLLLip** [aws-greengrass/aws-greengrass-nucleus#917]
**Describe the bug**
Tried to create a new version to deploy to one of my cores now that has been running for a couple of weeks, and the deployment just stands as in progress. Logged in to the core and checked the greengrass.log and it says something about failing to fetch credentials due to timeouts and broken pipe.

This issue appears in the logs today at 2021-04-09T03:01:00.218Z, and the previous log statement then was at 
`2021-03-29T23:50:00.700Z [INFO] (pool-2-thread-38) com.aws.greengrass.tes.CredentialRequestHandler: Received IAM credentials that will be cached until 2021-03-30T00:45:00Z. {iotCredentialsPath=/role-aliases/GreengrassV2TokenExchangeRoleAlias/credentials}
`

**To Reproduce**
Keep greengrass running for an extended period, then try to deploy.

**Expected behavior**
The application to recover, and deployment to succeed without manual intervention.

**Actual behavior**
Nothing happens part from logs full of errors.

**Environment**
 - OS: Raspbian Lite Buster
 - JDK version: openjdk version "11.0.9.1" 2020-11-04
OpenJDK Runtime Environment (build 11.0.9.1+1-post-Raspbian-1deb10u2)
OpenJDK Server VM (build 11.0.9.1+1-post-Raspbian-1deb10u2, mixed mode)
 - Nucleus version: 2.0.5

**Additional context**
The issue makes it impossible to deploy new versions to the core.

```
2021-04-09T13:01:00.355Z [ERROR] (pool-2-thread-5390) com.aws.greengrass.iot.IotConnectionManager: Getting connection failed for endpoint https://c31e8o2fzpui1i.credentials.iot.eu-central-1.amazonaws.com with error {} . {}
java.util.concurrent.TimeoutException
	at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1886)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2021)
	at com.aws.greengrass.iot.IotConnectionManager.getConnection(IotConnectionManager.java:111)
	at com.aws.greengrass.iot.IotCloudHelper.sendHttpRequest(IotCloudHelper.java:76)
	at com.aws.greengrass.tes.CredentialRequestHandler.getCredentialsBypassCache(CredentialRequestHandler.java:168)
	at com.aws.greengrass.tes.CredentialRequestHandler.getCredentials(CredentialRequestHandler.java:254)
	at com.aws.greengrass.tes.CredentialRequestHandler.handle(CredentialRequestHandler.java:133)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
	at jdk.httpserver/sun.net.httpserver.AuthFilter.doFilter(AuthFilter.java:82)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:80)
	at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle(ServerImpl.java:692)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
	at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange.run(ServerImpl.java:664)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

2021-04-09T13:01:00.358Z [WARN] (pool-2-thread-5390) com.aws.greengrass.tes.CredentialRequestHandler: Encountered error while fetching credentials. {iotCredentialsPath=/role-aliases/GreengrassV2TokenExchangeRoleAlias/credentials}
com.aws.greengrass.deployment.exceptions.AWSIotException: java.util.concurrent.TimeoutException
	at com.aws.greengrass.iot.IotConnectionManager.getConnection(IotConnectionManager.java:114)
	at com.aws.greengrass.iot.IotCloudHelper.sendHttpRequest(IotCloudHelper.java:76)
	at com.aws.greengrass.tes.CredentialRequestHandler.getCredentialsBypassCache(CredentialRequestHandler.java:168)
	at com.aws.greengrass.tes.CredentialRequestHandler.getCredentials(CredentialRequestHandler.java:254)
	at com.aws.greengrass.tes.CredentialRequestHandler.handle(CredentialRequestHandler.java:133)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
	at jdk.httpserver/sun.net.httpserver.AuthFilter.doFilter(AuthFilter.java:82)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:80)
	at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle(ServerImpl.java:692)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
	at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange.run(ServerImpl.java:664)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.util.concurrent.TimeoutException
	at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1886)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2021)
	at com.aws.greengrass.iot.IotConnectionManager.getConnection(IotConnectionManager.java:111)
	... 13 more

2021-04-09T13:01:00.361Z [INFO] (pool-2-thread-5390) com.aws.greengrass.tes.CredentialRequestHandler: Request failed. {}
java.io.IOException: Broken pipe
	at java.base/sun.nio.ch.FileDispatcherImpl.write0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:47)
	at java.base/sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:113)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:79)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:50)
	at java.base/sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:463)
	at jdk.httpserver/sun.net.httpserver.Request$WriteStream.write(Request.java:391)
	at jdk.httpserver/sun.net.httpserver.FixedLengthOutputStream.write(FixedLengthOutputStream.java:78)
	at java.base/java.io.FilterOutputStream.write(FilterOutputStream.java:108)
	at jdk.httpserver/sun.net.httpserver.PlaceholderOutputStream.write(ExchangeImpl.java:444)
	at com.aws.greengrass.tes.CredentialRequestHandler.handle(CredentialRequestHandler.java:135)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
	at jdk.httpserver/sun.net.httpserver.AuthFilter.doFilter(AuthFilter.java:82)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:80)
	at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle(ServerImpl.java:692)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
	at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange.run(ServerImpl.java:664)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by fiLLLip** [comment]
Additional info:
Restarting greengrass service with `systemctl restart greengrass` fixes the current issue and the deploy is successful, but it's not the first time I see this, and suspecting it's not the last...

**Comment by MikeDombo** [comment]
Hello Filip,
We have also seen this issue before and we are currently investigating it. Thanks for the report.

**Comment by MikeDombo** [comment]
Filip,
Can we get you to enable CRT logging by merging these JVM options in the Nucleus's configuration?
```json
{
  "jvmOptions": "-Daws.crt.log.destination=File -Daws.crt.log.filename=/tmp/crt.log -Daws.crt.log.level=Trace"
}
```

**Comment by fiLLLip** [comment]
> Filip,
> Can we get you to enable CRT logging by merging these JVM options in the Nucleus's configuration?
> 
> ```json
> {
>   "jvmOptions": "-Daws.crt.log.destination=File -Daws.crt.log.filename=/tmp/crt.log -Daws.crt.log.level=Trace"
> }
> ```

I have activated it now. I'll get back to you if it fails with the same errors again.

**Comment by MikeDombo** [comment]
Fixed in PR #918 by removing our usage of the CRT and instead using the AWS SDK client. This fix should be included in our upcoming release.

Please do get back to us if it reproduces with the CRT logs enabled, we'd still be interesting in sharing those findings with the SDK team.


Closing for now since the fix has been committed.

**Comment by fiLLLip** [comment]
@MikeDombo I could provide you with my /tmp/crt.log from the last period, if you still want it?

Do you know when this fix will make it into production? We are having serious issues caused by this.

**Comment by MikeDombo** [comment]
Hi,
I'm sorry but I cannot share any date at this time except to say that it should be soon. Since Greengrass is open source, you can clone this repository and build it using `mvn package -DskipTests` and then create a private version of the Nucleus in your account.

**Comment by MikeDombo** [comment]
The fix has been released in Nucleus version 2.1.0 which is now available globally.
